### PR TITLE
[SPARK-9397] DataFrame should provide an API to find source data files if applicable

### PR DESCRIPTION
**Pull request by aarondav**
Certain applications would benefit from being able to inspect DataFrames that are straightforwardly produced by data sources that stem from files, and find out their source data. For example, one might want to display to a user the size of the data underlying a table, or to copy or mutate it.

This PR exposes an `inputFiles` method on DataFrame which attempts to discover the source data in a best-effort manner, by inspecting HadoopFsRelations and JSONRelations.

### CODE REVIEW

**Comment by aarondav** [comment]
cc @marmbrus @rxin about exposing a new DataFrame API.

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 1]
```diff
@@ -1546,6 +1547,21 @@ class DataFrame private[sql](
     }
   }
 
+  /**
+   * Returns a best-effort snapshot of the files that compose this DataFrame. This method simply
+   * asks each constituent BaseRelation for its respective files and takes the union of all results.
+   * Depending on the source relations, this may not find all input paths. Duplicates are removed.
+   */
+  def sourcePaths: Array[String] = {
```
Comment (rxin) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: how about inputFiles?

**Comment by aarondav** [comment]
Renamed to inputFiles and updated description.

**Comment by rxin** [comment]
The API change LGTM.

**Pull request closed without merging**